xbmc/pvr/channels/PVRChannelProviders.cpp

```diff
@@ -0,0 +1,370 @@
+
+      (*it)->DeleteFromDatabase();
+      it = m_providers.erase(it);
+
+      bChanged |= true;
+    }
+    else
+    {
+      ++it;
+    }
+  }
+
+  return bChanged;
+}
+
+bool CPVRChannelProviders::UpdateClientEntries(const CPVRChannelProvidersContainer& newProviders,
+                                               const std::vector<int>& failedClients)
+{
+  bool bChanged = false;
+
+  CSingleLock lock(m_critSection);
+
+  /* go through the provider list and check for updated or new providers */
+  for (const auto& newProvider : newProviders.GetProvidersList())
+  {
+    bChanged |= !!CheckAndPersistEntry(newProvider, ProviderUpdateMode::BY_CLIENT);
+  }
+
+  /* check for deleted providers */
+  for (std::vector<std::shared_ptr<CPVRChannelProvider>>::iterator it = m_providers.begin();
+       it != m_providers.end();)
+  {
+    const std::shared_ptr<CPVRChannelProvider> provider = *it;
+    if (!newProviders.GetByClient(provider->GetClientId(), provider->GetUniqueId()))
+    {
+      /* provider was not found */
+      bool bIgnoreProvider = false;
+      for (const auto& failedClient : failedClients)
+      {
+        if (failedClient == provider->GetClientId())
+        {
+          bIgnoreProvider = true;
+          break;
+        }
+      }
+
+      /* ignore add-on providers as they are a special case */
+      if (provider->IsClientProvider())
+        bIgnoreProvider = true;
+
+      if (bIgnoreProvider)
+      {
+        ++it;
+        continue;
+      }
+
+      CLog::LogFC(LOGDEBUG, LOGPVR, "Deleted provider {} on client {}", provider->GetUniqueId(),
+                  provider->GetClientId());
+
+      (*it)->DeleteFromDatabase();
+      it = m_providers.erase(it);
+
+      bChanged = true;
+    }
+    else
+    {
+      ++it;
+    }
+  }
+
+  m_failedClients = failedClients;
+  m_bIsUpdating = false;
+
+  return bChanged;
+}
+
+std::shared_ptr<CPVRChannelProvider> CPVRChannelProviders::CheckAndAddEntry(
+    const std::shared_ptr<CPVRChannelProvider>& newProvider, ProviderUpdateMode updateMode)
+{
+  bool bChanged = false;
+
+  CSingleLock lock(m_critSection);
+  std::shared_ptr<CPVRChannelProvider> provider =
+      GetByClient(newProvider->GetClientId(), newProvider->GetUniqueId());
+  if (provider)
+  {
+    bChanged = provider->UpdateEntry(newProvider, updateMode);
+  }
+  else
+  {
+    provider = newProvider;
+    // We don't set and id as this came from the DB so it has one already
+    InsertEntry(provider, updateMode);
+
+    if (provider->GetProviderId() > m_iLastId)
+      m_iLastId = provider->GetProviderId();
+
+    bChanged = true;
+  }
+
+  return bChanged ? provider : std::shared_ptr<CPVRChannelProvider>();
+}
+
+std::shared_ptr<CPVRChannelProvider> CPVRChannelProviders::CheckAndPersistEntry(
+    const std::shared_ptr<CPVRChannelProvider>& newProvider, ProviderUpdateMode updateMode)
+{
+  bool bChanged = false;
+
+  CSingleLock lock(m_critSection);
+  std::shared_ptr<CPVRChannelProvider> provider =
```

Comment:
`const std::shared...`

Reply:
Not this time. provider is in the else condition.